# Laboratorio 4
### Laboratorio: Mejorando el Análisis de Sentimientos con LSTM y Características Adicionales
* Objetivo: Incrementar la precisión en el análisis de sentimientos sobre las críticas de películas
utilizando RNNs con unidades LSTM y la incorporación de características (features) adicionales.

# Integrantes

* Andrea Lam
* Gabriel Vicente

# 1. Importación y Pre-procesamiento (30 puntos)
• Correcta importación del conjunto de datos con 50,000 palabras más frecuentes:
10 puntos. <br/>
• Secuenciación y relleno de las críticas: 10 puntos.<br/>
• Extracción y adecuada justificación de características adicionales: 10 puntos.<br/>

In [2]:
!pip install tensorflow

  Obtaining dependency information for tensorflow from https://files.pythonhosted.org/packages/1b/66/2f47c39cfedb29188d82555d0184a619a0bf8234fd5e5301940efb0aa464/tensorflow-2.13.0-cp39-cp39-win_amd64.whl.metadata
  Using cached tensorflow-2.13.0-cp39-cp39-win_amd64.whl.metadata (2.6 kB)
  Obtaining dependency information for tensorflow-intel==2.13.0 from https://files.pythonhosted.org/packages/2b/ad/d3a2e335004d178e0599cf8aff6c2a92cd21eb9789358fb8f3f951009930/tensorflow_intel-2.13.0-cp39-cp39-win_amd64.whl.metadata
  Using cached tensorflow_intel-2.13.0-cp39-cp39-win_amd64.whl.metadata (4.1 kB)
  Obtaining dependency information for absl-py>=1.0.0 from https://files.pythonhosted.org/packages/01/e4/dc0a1dcc4e74e08d7abedab278c795eef54a224363bb18f5692f416d834f/absl_py-2.0.0-py3-none-any.whl.metadata
  Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
  Obtaining dependency information for flatbuffers>=23.1.21 from https://files.pythonhosted.org/packages/6f/12/d5c79ee252793ffe845d

In [3]:
""" Tensor flow """

import tensorflow as tf
from tensorflow.keras.preprocessing import sequence
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Embedding
from tensorflow.keras.layers import LSTM
from tensorflow.keras.datasets import imdb

In [4]:
print('Cargando los datos...')
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=50000)

Cargando los datos...
17464789/17464789 [==============================] - 2s 0us/step


In [5]:
X_train = sequence.pad_sequences(X_train, maxlen = 80)
X_test = sequence.pad_sequences(X_test, maxlen = 80)

# 2. Arquitectura del Modelo (30 puntos)
• Adecuado diseño de la arquitectura LSTM: 10 puntos.<br/>
• Incorporación efectiva de características adicionales en el modelo: 10 puntos.<br/>
• Uso de técnicas adicionales (e.g., Dropout, capas densamente conectadas): 10
puntos.

modelo original

In [19]:
modelo_original = Sequential()
modelo_original.add(Embedding(50000, 128))
modelo_original.add(LSTM(128, dropout=0.2, recurrent_dropout=0.2))
modelo_original.add(Dense(1, activation='sigmoid'))
modelo_original.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

modelo con mejoras

In [13]:
from keras.models import Sequential
from keras.layers import Embedding, LSTM, Dense, Dropout, Conv1D, MaxPooling1D

modelo = Sequential()
modelo.add(Embedding(50000, 128, input_length=80))  # Asegúrate de especificar input_length
modelo.add(Dropout(0.2))  # Regularización Dropout en la capa de embedding

# Capas Convolucionales
modelo.add(Conv1D(64, 5, activation='relu'))
modelo.add(MaxPooling1D(pool_size=4))
modelo.add(Dropout(0.2))  # Regularización Dropout después de las capas convolucionales

# Capas LSTM
modelo.add(LSTM(128, dropout=0.2, recurrent_dropout=0.2, return_sequences=True))
modelo.add(Dropout(0.2))
modelo.add(LSTM(128, dropout=0.2, recurrent_dropout=0.2))
modelo.add(Dropout(0.2))

# Capa de salida
modelo.add(Dense(1, activation='sigmoid'))

modelo.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# 3. Entrenamiento y Evaluación (20 puntos)
• Correcto entrenamiento del modelo sin errores: 10 puntos. <br/>
• Evaluación y comparación adecuada con el modelo del ejercicio anterior: 10
puntos.

modelo con cambios

In [14]:
modelo.fit(X_train, y_train,batch_size = 32,epochs = 15,verbose = 2,validation_data = (X_test, y_test))

Epoch 1/15
782/782 - 99s - loss: 0.4218 - accuracy: 0.7992 - val_loss: 0.3412 - val_accuracy: 0.8486 - 99s/epoch - 126ms/step
Epoch 2/15
782/782 - 97s - loss: 0.2089 - accuracy: 0.9201 - val_loss: 0.4141 - val_accuracy: 0.8346 - 97s/epoch - 124ms/step
Epoch 3/15
782/782 - 99s - loss: 0.0925 - accuracy: 0.9671 - val_loss: 0.4574 - val_accuracy: 0.8310 - 99s/epoch - 127ms/step
Epoch 4/15
782/782 - 100s - loss: 0.0429 - accuracy: 0.9857 - val_loss: 0.6319 - val_accuracy: 0.8147 - 100s/epoch - 128ms/step
Epoch 5/15
782/782 - 101s - loss: 0.0242 - accuracy: 0.9920 - val_loss: 0.7781 - val_accuracy: 0.8140 - 101s/epoch - 129ms/step
Epoch 6/15
782/782 - 102s - loss: 0.0174 - accuracy: 0.9939 - val_loss: 0.9336 - val_accuracy: 0.8129 - 102s/epoch - 131ms/step
Epoch 7/15
782/782 - 101s - loss: 0.0161 - accuracy: 0.9944 - val_loss: 1.0608 - val_accuracy: 0.8076 - 101s/epoch - 130ms/step
Epoch 8/15
782/782 - 102s - loss: 0.0122 - accuracy: 0.9953 - val_loss: 1.0909 - val_accuracy: 0.8119 - 102s/e

modelo original

In [21]:
modelo_original.fit(X_train, y_train,batch_size = 32,epochs = 15,verbose = 2,validation_data = (X_test, y_test))

Epoch 1/15
782/782 - 140s - loss: 0.3359 - accuracy: 0.8576 - val_loss: 0.3989 - val_accuracy: 0.8208 - 140s/epoch - 180ms/step
Epoch 2/15
782/782 - 139s - loss: 0.1902 - accuracy: 0.9276 - val_loss: 0.4003 - val_accuracy: 0.8291 - 139s/epoch - 178ms/step
Epoch 3/15
782/782 - 140s - loss: 0.1087 - accuracy: 0.9608 - val_loss: 0.5350 - val_accuracy: 0.8223 - 140s/epoch - 180ms/step
Epoch 4/15
782/782 - 140s - loss: 0.0604 - accuracy: 0.9796 - val_loss: 0.6311 - val_accuracy: 0.8126 - 140s/epoch - 179ms/step
Epoch 5/15
782/782 - 140s - loss: 0.0510 - accuracy: 0.9826 - val_loss: 0.6698 - val_accuracy: 0.8051 - 140s/epoch - 179ms/step
Epoch 6/15
782/782 - 143s - loss: 0.0328 - accuracy: 0.9894 - val_loss: 0.9203 - val_accuracy: 0.8107 - 143s/epoch - 183ms/step
Epoch 7/15
782/782 - 143s - loss: 0.0234 - accuracy: 0.9926 - val_loss: 0.8296 - val_accuracy: 0.8128 - 143s/epoch - 183ms/step
Epoch 8/15
782/782 - 140s - loss: 0.0156 - accuracy: 0.9950 - val_loss: 0.9951 - val_accuracy: 0.8115 - 

modelo con cambios

In [15]:
from tabulate import tabulate

loss, accuracy = modelo.evaluate(X_test, y_test, batch_size=32, verbose=2)

# Definir los títulos y resultados
headers = ["Métrica", "Resultado"]
data = [["Loss", f"{loss:.4f}"],
        ["Test Accuracy", f"{accuracy * 100:.2f}%"]]

# Imprimir la tabla utilizando tabulate
table = tabulate(data, headers, tablefmt="fancy_grid")
print(table)


782/782 - 7s - loss: 1.2729 - accuracy: 0.8090 - 7s/epoch - 9ms/step
╒═══════════════╤═════════════╕
│ Métrica       │ Resultado   │
╞═══════════════╪═════════════╡
│ Loss          │ 1.2729      │
├───────────────┼─────────────┤
│ Test Accuracy │ 80.90%      │
╘═══════════════╧═════════════╛


modelo original

In [22]:
from tabulate import tabulate

loss, accuracy = modelo_original.evaluate(X_test, y_test, batch_size=32, verbose=2)

# Definir los títulos y resultados
headers = ["Métrica", "Resultado"]
data = [["Loss", f"{loss:.4f}"],
        ["Test Accuracy", f"{accuracy * 100:.2f}%"]]

# Imprimir la tabla utilizando tabulate
table = tabulate(data, headers, tablefmt="fancy_grid")
print(table)


782/782 - 10s - loss: 1.2061 - accuracy: 0.8023 - 10s/epoch - 12ms/step
╒═══════════════╤═════════════╕
│ Métrica       │ Resultado   │
╞═══════════════╪═════════════╡
│ Loss          │ 1.2061      │
├───────────────┼─────────────┤
│ Test Accuracy │ 80.23%      │
╘═══════════════╧═════════════╛


In [ ]:
modelo.save("Lab4.h5")

# 4. Informe

**Clara descripción de características adicionales y su relevancia: 5 puntos.**

Se agregaron capas de Dropout después de la capa de embedding, después de las capas convolucionales y después de las capas LSTM para aplicar regularización y reducir el sobreajuste.

Se añadieron capas CNN para capturar patrones locales en el texto antes de las capas LSTM.

**Explicación coherente de la arquitectura del modelo: 5 puntos.**

El modelo está configurado para minimizar la función de pérdida de entropía cruzada binaria ('binary_crossentropy') utilizando el optimizador 'adam', y evalúa su rendimiento en términos de precisión ('accuracy'). Esta arquitectura combina capas de embedding, convolucionales y LSTM para procesar secuencias de texto y aprender representaciones significativas de las palabras, lo que permite la clasificación binaria de los sentimientos en textos. La regularización y las capas Dropout ayudan a prevenir el sobreajuste durante el entrenamiento.

**Presentación de resultados y comparativa efectiva: 10 puntos.**

Como se puede observar en el modelo original se tuvo un menor loss del 1.2061, mientras que el loss del modelo mejora fue de 1.2729; esto se puede deber a diferentes razones como la arquitectura del modelo, los hiperparametros, la inilicacion de pesos o la regularización.
Pero por otro lado el accuracy fue mejor para el modelo mejorado ya que obtuvo un 80.90% mientras el modelo original tiene un 80.23%